Added CSS Tricks (#112)
xsleaks/wiki: content/docs/attacks/css-tricks.md

## Evil Captcha
Using CSS, it’s possible to take an embed out of context.  
An example of this is pretending it’s a captcha. [source](https://ronmasas.com/posts/the-human-side-channel)  
This can be prevented with XFO.

## Abusing autocomplete
If a website uses [autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) it may be possible to leak data such as email addresses.
This does require the user to be tricked into pressing the Up or Down arrow key then pressing Enter or Tab.
```javascript
let input = document.createElement("input");
input.type = "email";
input.autocomplete = "email";
input.name = "email";
input.size = "1";
input.style = "position:absolute;right:-500px;bottom:-21.9px";
input.onkeypress = e => {
    e.preventDefault();
}
window.onmousedown = e => {
    // ignore mouse clicks
    e.preventDefault();
}
input.onchange = e => {
    alert(e.srcElement.value);
    e.srcElement.value = "";
}
document.body.appendChild(input);
setInterval(() => {
    input.focus({preventScroll: true});
}, 1000);
```

## Custom cursor
Might not leak data directly but it may help trick the user.
```css
:root {
   cursor: url(URL), auto;            
}
```

> ## Defence

Reviewer: ```suggestion
## Defense
```

## Getting history
A PoC of this attack can be found [^leak-1].  
Using the CSS [`:visited`](https://developer.mozilla.org/en-US/docs/Web/CSS/:visited) selector, it’s possible to apply a different style for URLs that have been visited.  
Previously it was possible to use [`getComputedStyle()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) to detect this difference but now browsers prevent this. [^changes-1]   
So, it may be needed to trick the user into clicking an area that the CSS has affected this can be done using [`mix-blend-mode`](https://developer.mozilla.org/en-US/docs/Web/CSS/mix-blend-mode). [^blend-mode]   
Render timings have also been abused to do this without tricking the user. [^render-timings]  

## Evil Captcha
Using CSS, it’s possible to take an embed out of context.  
An example of this is pretending it’s a captcha. [^leak-2]
This can be prevented with XFO.

## Abusing autocomplete
If a website uses [autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) it may be possible to leak data such as email addresses.

> This does require the user to be tricked into pressing the Up or Down arrow key then pressing Enter or Tab.

Reviewer: I think the construction of the sentence is confusing. As I understand: "It is possible to leak user's data, such as email addresses, that was typed into `<input>` elements that have `<autocomplete>` specified."

Plus, isn't this the default? Doesn't it work the other way around that you have to specify no-autocomplete to disable it? Also, worth specifying that it only affect text inputs.

> This does require the user to be tricked into pressing the Up or Down arrow key then pressing Enter or Tab.

Reviewer: Needs explanation why. It's to iterate through different suggestions right?

Author: Yes pressing the Up or Down selects a option from the always focused input then Enter or Tab submits it.

```javascript
let input = document.createElement("input");
input.type = "email";
input.autocomplete = "email";
input.name = "email";
input.size = "1";
input.style = "position:absolute;right:-500px;bottom:-21.9px";
input.onkeypress = e => {
    e.preventDefault();
}
window.onmousedown = e => {
    // ignore mouse clicks
    e.preventDefault();
}
input.onchange = e => {
    alert(e.srcElement.value);
    e.srcElement.value = "";
}
document.body.appendChild(input);
setInterval(() => {
    input.focus({preventScroll: true});
}, 1000);
```

## Custom cursor

> Might not leak data directly but it may help trick the user.

Reviewer: I don't understand this paragraph. How can this trick the user? Into what?

Author: https://microsoftedge.github.io/edgevr/posts/ui-security-thinking-outside-the-viewport/#cve-2020-15985-cursor-hijacking-mitigation-bypass

```css
:root {
   cursor: url(URL), auto;            
}
```

## Defense

> Some of the attacks can be prevented by using XFO,

Reviewer: Can you expand how and specify which attacks? I don't immediately see how this prevents and which attacks.

> The history attack can be prevented by disabling history or if on Firefox it’s possible to fix this by setting the `layout.css.visited_links_enabled` option in `about:config` to `false`.  

Reviewer: ```suggestion
The *Retrieving user's history* attack can be only prevented by the user. It can be done by disabling the browser history, or if on Firefox, by setting the option `layout.css.visited_links_enabled` to `false` in `about:config` panel.  
```

Number of bugs were reported about the issue: [^bug-1], [^bug-2], [^bug-3]
| [SameSite Cookies (Lax)]({{< ref "/docs/defenses/opt-in/same-site-cookies.md" >}}) | [COOP]({{< ref "/docs/defenses/opt-in/coop.md" >}}) | [Framing Protections]({{< ref "/docs/defenses/opt-in/xfo.md" >}}) |                  [Isolation Policies]({{< ref "/docs/defenses/isolation-policies" >}})                   |
| :--------------------------------------------------------------------------------: | :-------------------------------------------------: | :---------------------------------------------------------------: | :------------------------------------------------------------------------------------------------------: |

> |                                         ❌                                          |                          ❌                          |                                 ✔️                                 |  |

Reviewer: ```suggestion
|                                         ❌                                          |                          ❌                          |                                 ✔️                                 | ❌  |
```

CSS can be used to trick a user into exposing information such as embedded pixel values by making visual changes that are affected by the embed.

> ## Getting history

Reviewer: ```suggestion
## Retrieving user's history
```

> A PoC of this attack can be found [^leak-1].  

Reviewer: The sentence doesn't make much sense after adding the reference. This specific case is good to be referenced directly in the text. Though, the PoC is outdated right? I would replace all bullet points to a paragraph. Then it will come naturally to add references.

Author: PoC still works for me.

Author: Im not aware of it being outdated.

Using the CSS [`:visited`](https://developer.mozilla.org/en-US/docs/Web/CSS/:visited) selector, it’s possible to apply a different style for URLs that have been visited.  
Previously it was possible to use [`getComputedStyle()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) to detect this difference but now browsers prevent this. [^changes-1]   
So, it may be needed to trick the user into clicking an area that the CSS has affected this can be done using [`mix-blend-mode`](https://developer.mozilla.org/en-US/docs/Web/CSS/mix-blend-mode). [^blend-mode]   
Render timings have also been abused to do this without tricking the user. [^render-timings]  

## Evil Captcha
Using CSS, it’s possible to take an embed out of context.  

> An example of this is pretending it’s a captcha. [^leak-2]

Reviewer: Same as before, the sentence doesn't make much sense now. It needs more description that will justify adding a reference to it.

Author: Does not help [^leak-2] is not working for me now.

}
```

## Defense
Some of the attacks can be prevented by using XFO,
The history attack can be prevented by disabling history or if on Firefox it’s possible to fix this by setting the `layout.css.visited_links_enabled` option in `about:config` to `false`.  

> Number of bugs were reported about the issue: [^bug-1], [^bug-2], [^bug-3]

Reviewer: References are to add additional information for readers, or justify the statement.

`Number of bugs about leaking user's history were reported to browsers. [^bug-1] [^bug-2] [^bug-3]`, also, this doesn't seem to fit the defense section since the bugs are about leaking user's history through `:visited` I would add it as `<{{ hint info }}>` *Getting history* section.

Author: Moved.
